src/frontends/ir/src/ir_deserializer.cpp

```diff
@@ -346,11 +346,8 @@ void XmlDeserializer::on_adapter(const std::string& name, ngraph::ValueAccessor<
                 IE_THROW() << "Attribute and shape size are inconsistent for " << type << " op!";
 
             char* data = m_weights->get_ptr<char>() + offset;
-            auto buffer =
-                std::make_shared<ngraph::runtime::SharedBuffer<std::shared_ptr<ngraph::runtime::AlignedBuffer>>>(
-                    data,
-                    size,
-                    m_weights);
+            auto buffer = std::make_shared<ngraph::runtime::AlignedBuffer>(size);
```

Comment:
Why do we need to have this change? Here you increase the memory consumption in 2 times.

Reply:
Current design is like we load whole bin file into a big buffer (call it A) and all constant nodes (const1,const2,... constN) create `SharedBuffer` referencing part of it, thus buffer A is referenced N times, this is bad when ngraph const-folding pass is folding a constant operation into another constant,  for example, `const_new1= mul(const1,32)`,  after folding, memory taken by const1 should be freed with the const1 node itself, but with `SharedBuffer`, it's not, it only decrease reference count of A by 1, since not all const is folded, the big buffer A still will be referenced by some const node and cannot be partially freed.

Thus constant folding pass will increase total memory consumption(sometimes by a lot).

Here we try to break the big buffer A into smaller ones that has it's own life-cycle, offcause this may temporarily double the memory consumption since the big buffer A is co-exist when during deserialization, but after all model is desterilized and no constant node will be referencing buffer A and A will be eventually freed.

maybe we can take one more step to eliminate the needs of buffer A totally, so the peak memory usage won;t be effected, shall we?

Reply:
But let's imagine another situation. We have a model with a lot of different constant operations which uses the same memory. In current approach you will duplicate this memory several times and it also will increase memory consumption.

In the majority of situations constant folding doesn't affect weights of convolution or MatMul which constitute the major impact for the weights size.

Before we also done some experiments for reading from bin file for each constant, but it have an increase of the reading time (because we spend a lot of time for jumping in the bin file and reading weights).

At the current moment the approach with weights sharing looks like the best compromise between memory consumption and read time performance.

Moreover I don't remember any tickets from customers which complain about memory consumption after the constant folding.

Reply:
Your concern is reasonable, it's a mbart model from ticket CVS-82367, the constant folding happens on Embedding (Gather) rather than convolution and it's huge. any suggestion on better solution?


Result_155916 torch-jit-export(
	input_ids,
	attention_mask,
) {
	i32_{?,?}  attention_mask = Parameter()
	i32_{?,?}  input_ids = Parameter()
	f32_{250054,1024}  **embed_tokens.weight** = Constant(..., Allocated)
	f32_{}  Constant_267256 = Constant(32, Allocated)
	f32_{250054,1024}  **Multiply_267257** = Multiply(**embed_tokens.weight,Constant_267256**)
	i32_{2}  206 = Constant(-1,0, Allocated)
	i32_{?,?}  207 = Reshape(input_ids,206)
	i32_{}  Constant_454 = Constant(0, Allocated)
	f32_{?,?,1024}  208 = Gather(**Multiply_267257,207**,Constant_454)